### PR TITLE
Integration test for `CachingSession`

### PR DESCRIPTION
**Pull request by Bouncheck**
Adds the integration test `caching_session::ensure_cache_is_used` for `CachingSession`. The test sends through the `scylla_proxy` several queries and verifies that the number of prepare requests is exactly as expected. If the cache is working correctly no extra prepare requests should appear. If it's not, the assertions will bring attention to that.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
*No public items introduced*
- [x] I have adjusted the documentation in `./docs/source/`.
*No docs to change for the test. Docs for CachingSession are separate issue.*
- [x] I added appropriate `Fixes:` annotations to PR description.
*No open issues regarding integration tests for CachingSession*

### CODE REVIEW

File: scylla/tests/integration/caching_session.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            let ks = unique_keyspace_name();
+
+            let (feedback_txs, mut feedback_rxs): (Vec<_>, Vec<_>) = (0..3)
+                .map(|_| mpsc::unbounded_channel::<(RequestFrame, Option<TargetShard>)>())
+                .unzip();
+            for (i, tx) in feedback_txs.iter().cloned().enumerate() {
+                running_proxy.running_nodes[i].change_request_rules(Some(vec![RequestRule(
+                    Condition::and(Condition::RequestOpcode(RequestOpcode::Prepare),Condition::not(Condition::ConnectionRegisteredAnyEvent)),
+                    RequestReaction::noop().with_feedback_when_performed(tx),
+                )]));
```
Comment (wprzytula) [inline scylla/tests/integration/caching_session.rs:45]: :wrench: This doesn't seem like formatted correctly, even though the format check has passed...
Please format this manually, as it's a pain to read.

File: scylla/tests/integration/caching_session.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            }
+
+
+            session.ddl(format!("CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1}}", ks)).await.unwrap();
```
Comment (wprzytula) [inline scylla/tests/integration/caching_session.rs:49]: :wrench: ditto: formatting here.

File: scylla/tests/integration/caching_session.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            session.use_keyspace(ks, false).await.unwrap();
+            session
```
Comment (Lorak-mmk) [inline scylla/tests/integration/caching_session.rs:51]: Could you split this string into two lines? Here you are running into a nasty edge case in rustfmt, where if a line length exceeds its limit, and its not possible to make it shorted (e.g. a too long string literal) it silently gives up formatting the whole function. This causes your whole test to not be formatted properly.

File: scylla/tests/integration/caching_session.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                .ddl("CREATE TABLE IF NOT EXISTS tab (a int, b int, c int, primary key (a, b, c))")
+                .await
+                .unwrap();
+            // Assumption: all nodes have the same number of shards
+            let nr_shards = match session.get_cluster_state().get_nodes_info().first() {
+                Some(node_info) => match node_info.sharder() {
+                    Some(sharder) => sharder.nr_shards,
+                    None => NonZero::new(1).unwrap(), // If there is no sharder, assume 1 shard
+                },
+                None => panic!("No nodes information available"),
```
Comment (wprzytula) [inline scylla/tests/integration/caching_session.rs:61]: :pick:
```suggestion
            let nr_shards = session.get_cluster_state()
                .get_nodes_info()
                .first()
                .expect("No nodes information available")
                .map(|sharder| sharder.nr_shards)
                .unwrap_or(NonZero::new(1).unwrap()); // If there is no sharder, assume 1 shard.
```

File: scylla/tests/integration/caching_session.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            prepared_batch_res_rows.sort();
+            let expected_rows = vec![(1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4)];
+            assert_eq!(prepared_batch_res_rows, expected_rows);
+
+
+            let feedbacks: usize = feedback_rxs.iter_mut().map(consume_current_feedbacks).sum();
+
+            let expected_feedbacks = (4 + 1) * nr_shards.get() as usize * 3; // 4 inserts + 1 select, 3 nodes
+            assert_eq!(feedbacks, expected_feedbacks);
+
```
Comment (Lorak-mmk) [inline scylla/tests/integration/caching_session.rs:109]: Can you verify the feedbacks amount after first execution of the batch, and then again after other executions to verify there are no new prepares?

File: scylla/tests/integration/caching_session.rs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            batch.append_statement(insert_3_b_c);
+            batch.append_statement(insert_4_b_c);
+            // First batch that should generate prepares for each shard.
+            let batch_values = vec![(1, 1), (2, 2), (3, 3), (4, 4)];
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            // Few extra runs. Those batches should not result in any prepares being sent.
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
```
Comment (wprzytula) [inline scylla/tests/integration/caching_session.rs:84]: :pick: Let's use `for` loop instead of repetition here.

File: scylla/tests/integration/caching_session.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            let prepared_batch_res_rows: Vec<(i32, i32, i32)> = caching_session
+                .execute_unpaged("SELECT * FROM tab", &[])
+                .await
+                .unwrap()
+                .into_rows_result()
+                .unwrap()
+                .rows()
+                .unwrap()
+                .map(|r| r.unwrap())
+                .collect();
```
Comment (wprzytula) [inline scylla/tests/integration/caching_session.rs:95]: :pick: 
```suggestion
                .collect::<Result<_, _>>()
                .unwrap();
```

File: scylla/tests/integration/caching_session.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            // Consume those frames.
+            feedback_rxs.iter_mut().map(consume_current_feedbacks).sum::<usize>();
+
+            // Run batches many times per node to ensure we hit every one of them.
+            // Server side cache was invalidated, but the reprepare will be required only if
+            // we hit the node that had its cache invalidated. Only the corresponding shards
+            // will be targeted, so this time expected number of prepares is not multiplied by nr_shards
+            for _ in 0..(64*3) {
+                caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            }
```
Comment (Lorak-mmk) [inline scylla/tests/integration/caching_session.rs:126]: Instead of making a lot of queries (which can actually fail to reach all shards if you are unlucky) you can use a custom load balancing policy to send a query to each shard exactly once. See tablets.rs integration tests and `SingleTargetLBP` inside it (it is used in `send_statement_everywhere` and `send_unprepared_query_everywhere` helper functions).

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]

**Review by wprzytula — APPROVED** [review]

File: scylla/tests/integration/caching_session.rs (hunk 1 above)
Comment (Lorak-mmk) [reply scylla/tests/integration/caching_session.rs:45]: As I mentioned in my comment, this is caused by rustfmt giving up on formatting due to a too long string literal. Breaking this literal up to two lines will fix the issue.

File: scylla/tests/integration/caching_session.rs (hunk 3 above)
Comment (Bouncheck) [reply scylla/tests/integration/caching_session.rs:51]: I've put replication settings in another variable. Looks like `cargo fmt --all` went further this time.
Changes will be visible once pushed.

File: scylla/tests/integration/caching_session.rs (hunk 5 above)
Comment (Bouncheck) [reply scylla/tests/integration/caching_session.rs:109]: I've added more frequent checks. Now the number of prepares is verified after each query. I've also put those numbers into descriptive variables like `batch_size`, `cluster_size`.

File: scylla/tests/integration/caching_session.rs (hunk 8 above)
Comment (Bouncheck) [reply scylla/tests/integration/caching_session.rs:126]: Applied same logic as in `send_statement_everywhere` but in a synchronous manner. Moved some code from `tablets.rs` to `common/utils.rs` in a separate commit to use it.

File: scylla/tests/integration/caching_session.rs (hunk 4 above)
Comment (Bouncheck) [reply scylla/tests/integration/caching_session.rs:61]: Applied, will be visible after push.

File: scylla/tests/integration/caching_session.rs (hunk 6 above)
Comment (Bouncheck) [reply scylla/tests/integration/caching_session.rs:84]: Switched to for loops from 1 to `=batch_size`

**Comment by Bouncheck** [comment]
Pushed a new version. Let me know if should I mark conversations as resolved or if it's on the reviewer if he agrees with the new version.

File: scylla/tests/common/utils.rs [hunk 9]
```diff
@@ -300,3 +300,31 @@ impl PerformDDL for CachingSession {
         self.execute_unpaged(query, &[]).await.map(|_| ())
     }
 }
+
+#[derive(Debug)]
+#[allow(dead_code)]
```
Comment (Bouncheck) [inline scylla/tests/common/utils.rs:305]: I don't understand why I had to add this line here even though I use this struct in my test. Should it not be not a dead_code because of that?

**Comment by Lorak-mmk** [comment]
> Pushed a new version. Let me know if should I mark conversations as resolved or if it's on the reviewer if he agrees with the new version.

In this repo usually the author of the PR marks them as resolved. Is that the best way? I have no idea, maybe the reviewer should do this - we can discuss it in the future.

**Review by muzarski — APPROVED** [review]
